[layouts] import from 2.x

elpaso opened the pull request:
## Import composer items from 2.x

Not yet completed

## TODO

- [x] import multiframes
- [x] connect uuids
- [x] connect to the layout loader

elpaso committed:
[layout] Converter: initial implementation

elpaso committed:
[layout] Converter: shapes

elpaso committed:
[layout] Converter: fix frame fill in shapes

elpaso committed:
Renamed some private vars to follow coding guidelines

elpaso committed:
Added optional argument to create reference layout images

elpaso committed:
Added test composer templates for import test

elpaso committed:
[layout] Import some more items: map, poly* and arrow

elpaso committed:
[layout] Convert scalebar and legent from 2.x

elpaso committed:
Removed some TODOs

elpaso committed:
[layout] Disable ref images generation and added docs

elpaso committed:
Ignore spell check for qgs project

elpaso committed:
Remove SIP annotation

elpaso committed:
[layout] connect converted to loader code

elpaso committed:
Sipify

elpaso committed:
Removed docs for private members

elpaso committed:
Updated test for imported compositions

elpaso committed:
Added diff images for tests

DelazJ commented on src/core/layout/qgscompositionconverter.cpp:
```diff
@@ -0,0 +1,1414 @@
+    newItems << layoutItem ;
+  }
+
+  // Polylin
```
Polyline (i know it's a comment but...)

DelazJ commented on src/core/qgsmultirenderchecker.h:
```diff
@@ -202,12 +202,15 @@ class CORE_EXPORT QgsLayoutChecker : public QgsMultiRenderChecker
      * The maximum number of allowable pixels differing from the reference image is
      * specified via the \a pixelDiff argument.
      *
+     * A reference image can be creates by setting \a createReferenceImage to true
```
created

DelazJ commented on tests/src/core/testqgscompositionconverter.cpp:
```diff
@@ -0,0 +1,427 @@
+    void importComposerTemplateShape();
+
+    /**
+     * Test import multiple ements from a composer template
```
elements

elpaso replied:
Thanks! Fixed.

DelazJ reviewed (commented):
some typos

elpaso committed:
Typos

nyalldawson commented on src/core/layout/qgscompositionconverter.h:
```diff
@@ -0,0 +1,212 @@
+    /**
+     * Composition data defined properties for different item types
+     */
+    enum DataDefinedProperty
```
Good idea putting this here - it means we could remove some of the deprecated ones from QgsLayoutObject and clean it up

nyalldawson commented on src/core/layout/qgscompositionconverter.h:
```diff
@@ -0,0 +1,212 @@
+     * \param project the QGIS project
+     * \return a QgsLayout instance
+     * \since QGIS 3.0
+     * \note Not available in Python bindings.
```
I don't think you need to explicitly list these "not available" notes when the whole class has been marked this way (pretty sure there's handling the doxygen test for this)

elpaso replied:
Yeah, I was in doubt ...

nyalldawson commented on src/core/layout/qgscompositionconverter.h:
```diff
@@ -0,0 +1,212 @@
+     * \since QGIS 3.0
+     * \note Not available in Python bindings.
+     */
+    static QgsLayout *createLayoutFromCompositionXml( const QDomElement &parentElement,
```
Given that this isn't in the python bindings, can you return a std::unique_ptr< QgsLayout > instead?

nyalldawson commented on tests/src/core/testqgscompositionconverter.cpp:
```diff
@@ -0,0 +1,427 @@
+  QCOMPARE( label->frameStrokeWidth().length(), 0.2 );
+  QCOMPARE( ( int )label->rotation(), 4 );
+
+  checkRenderedImage( layout, QTest::currentTestFunction(), 0 );
```
Can you please push up the font size on this label? The reference image is very small, and cross-platform rendering differences will make this test unstable.
Also need to make sure it uses the test reference font.

elpaso replied:
The diff images are required to pass the test on Travis due to font differences, I've searched everywhere for documentation about the "safe" font for rendering tests on Travis but I couldn't find anything, additionally, existing tests apparently use a mix of "MS" and "Ubuntu" and "Courier" and "DejaVu" fonts so I could't guess what it was the "safe" one.

The test reference font to use is .... ?

nyalldawson commented on src/core/layout/qgscompositionconverter.cpp:
```diff
@@ -0,0 +1,1414 @@
+QgsLayout *QgsCompositionConverter::createLayoutFromCompositionXml( const QDomElement &parentElement, QgsProject *project )
+{
+  initPropertyDefinitions();
+  QgsLayout *layout = new QgsLayout( project );
```
std::unique_ptr< QgsLayout > layout = qgis::make_unique< QgsLayout >( project );

nyalldawson commented on src/core/layout/qgscompositionconverter.cpp:
```diff
@@ -0,0 +1,1414 @@
+  QgsLayout *layout = new QgsLayout( project );
+  // Create pages
+  int pages = parentElement.attribute( QStringLiteral( "numPages" ) ).toInt( );
+  float paperHeight = parentElement.attribute( QStringLiteral( "paperHeight" ) ).toFloat( );
```
Think this should be double instead

nyalldawson commented on src/core/layout/qgscompositionconverter.cpp:
```diff
@@ -0,0 +1,1414 @@
+
+  QString name = parentElement.attribute( QStringLiteral( "name" ) );
+  layout->setName( name );
+  // TODO: check that it is always landscape
```
TODO can be removed - there's no need to set the orientation manually

nyalldawson commented on src/core/layout/qgscompositionconverter.cpp:
```diff
@@ -0,0 +1,1414 @@
+  {
+    if ( pasteInPlace )
+    {
+      QgsLayoutPoint posOnPage = QgsLayoutPoint::decodePoint( itemNode.toElement().attribute( QStringLiteral( "positionOnPage" ) ) );
```
I don't think this will work - positionOnPage wasn't present in composition xml

nyalldawson commented on src/core/layout/qgscompositionconverter.cpp:
```diff
@@ -0,0 +1,1414 @@
+
+    QgsCompositionConverter::readXml( layoutItem, composerItemElem );
+
+    // Frame color
```
Are these a TODO?

elpaso replied:
Yes, I took the enums and used them as TODOs, I can remove it.

nyalldawson commented on src/core/layout/qgscompositionconverter.cpp:
```diff
@@ -0,0 +1,1414 @@
+    //to shift the items by so that they are placed at this point
+    //First, calculate the minimum position from the xml
+    // TODO: check this!
+    QPointF minItemPos = layout->minPointFromXml( parentElement );
```
again, I don't think this will work with composer xml - you might need to copy the composition method for this and move it to a private member here.

nyalldawson commented on src/core/layout/qgscompositionconverter.cpp:
```diff
@@ -0,0 +1,1414 @@
+    if ( shapeStyleSymbol )
+      layoutItem->setSymbol( shapeStyleSymbol );
+  } /*
+  QDomElement shapeStyleSymbolElem = itemElem.firstChildElement( QStringLiteral( "symbol" ) );
```
remove?

nyalldawson commented on src/core/layout/qgscompositionconverter.cpp:
```diff
@@ -0,0 +1,1414 @@
+  layoutItem->mCacheInvalidated = true;
+
+  //overviews
+  layoutItem->mOverviewStack->readXml( itemElem, doc, context );
```
just confirming - did you double check that the grid/overview readXml methods work ok with composition xml? (from memory they haven't changed, but I don't trust my memory :P)

elpaso replied:
There is a test with an overview and it apparently works, I didn't explicitly test the grid though (unless it's in the main map test).

nyalldawson commented on src/core/layout/qgscompositionconverter.cpp:
```diff
@@ -0,0 +1,1414 @@
+    layoutItem->mGridStack->addGrid( mapGrid );
+  }
+
+  /* TODO: skip?
```
Yep - kill it. People can "bounce" their ancient projects via 2.18 if they need to open them.

nyalldawson commented on src/core/layout/qgscompositionconverter.cpp:
```diff
@@ -0,0 +1,1414 @@
+    }
+  }
+
+  if ( 
```
not required anymore, kill it

nyalldawson reviewed (commented):
Looking good!

Can you remove the _result_diff.png images? (They shouldn't be needed)

pblottiere commented on src/core/layout/qgscompositionconverter.cpp:
```diff
@@ -0,0 +1,1414 @@
+      fillRed = fillColorElem.attribute( QStringLiteral( "red" ) ).toDouble( &redOk );
+      fillGreen = fillColorElem.attribute( QStringLiteral( "green" ) ).toDouble( &greenOk );
+      fillBlue = fillColorElem.attribute( QStringLiteral( "blue" ) ).toDouble( &blueOk );
+      fillAlpha = fillColorElem.attribute( QStringLiteral( "alpha" ) ).toDouble( &alphaOk );
```
Is there a reason to use the `toDouble()` method for integer variables like `fillRed`, `fillGreen`, ...?

pblottiere commented on src/core/layout/qgscompositionconverter.cpp:
```diff
@@ -0,0 +1,1414 @@
+
+
+QgsLayout *QgsCompositionConverter::createLayoutFromCompositionXml( const QDomElement &parentElement, QgsProject *project )
+{
```
Do you think that an assert could be necessary on `* project`?

elpaso replied:
yes, safer

elpaso committed:
[layouts] Removed diff images and adapted tests

to use QGIS font

Also addressed comments in the PR

elpaso committed:
[test] Do not check the unique_ptr but get

Because std::unique_ptr is not implicitly convertible to bool
(why it works locally remains a mistery to me)

elpaso committed:
[layouts] test masks for rendering tests

nirvn commented:
@elpaso , I've just tested this PR, great job.

I've spotted one issue: the map setting of legend item isn't imported into the layout equivalent.

elpaso commented:
@nirvn thanks for the feedback! It's in the TODOs in this PR, "connect uuids" is about connecting the map IDs, that are now UUIDs instead of the integer map id that was used in the compositions.
I'm waiting for @nyalldawson input about that issue.

elpaso committed:
[layout] Fixed relative paths for test resources

elpaso committed:
[layout] Use uuids for map references in items

Implemented:

- picture
- legend
- overview
- scalebar

elpaso committed:
[layout] Added tests for uuid linked maps

Also updated the control images

elpaso committed:
[layout] Check for pointer as bool instead of compare

Travis does not like the second.

elpaso committed:
[layout] Add map rendering to tests

elpaso committed:
[layout] Fix title from name when converting a composition

elpaso committed:
[layout] Import HTML and atlas from 2.x composition

elpaso committed:
[layout] Import attribute table from composition

elpaso committed:
Add print template to spell ignore

elpaso committed:
Fix test after composition bindings removal

nyalldawson commented on src/core/layout/qgscompositionconverter.h:
```diff
@@ -0,0 +1,231 @@
+     * createLayoutFromCompositionXml is a factory that creates layout instances from a
+     *  QGIS 2.x XML composition \a parentElement and a QGIS \a project
+     * \return a QgsPrintLayout instance
+     * \since QGIS 3.0
```
I think convention is to leave out the since annotations when it matches the class' version added.

nyalldawson commented on src/core/layout/qgslayoutmanager.cpp:
```diff
@@ -188,7 +189,22 @@ bool QgsLayoutManager::readXml( const QDomElement &element, const QDomDocument &
+      }
+    }
+
+    // legacy import (to be removed!)
```
Kill it now I think

nyalldawson commented on tests/src/core/testqgscompositionconverter.cpp:
```diff
@@ -0,0 +1,669 @@
+  const QgsLayoutItemShape *shape = items.at( 0 );
+  QCOMPARE( shape->pos().x(), 261.132 );
+  QCOMPARE( shape->pos().y(), 83.1791 );
+  QCOMPARE( shape->sizeWithUnits().width(), 12.0988 );
```
Can you also check here positionWithUnits too? (Test that unit is mm)

nyalldawson commented on src/core/layout/qgscompositionconverter.cpp:
```diff
@@ -0,0 +1,1685 @@
+
+  // look for stored layer name
+  QString layerId = itemElem.attribute( QStringLiteral( "coverageLayer" ) );
+  QString layerName = itemElem.attribute( QStri
```
Not needed, drop this

nyalldawson replied:
It's now a custom property on the layout - see https://github.com/qgis/QGIS/blob/master/src/app/layout/qgslayoutdesignerdialog.cpp#L2633

nyalldawson commented on src/core/layout/qgscompositionconverter.cpp:
```diff
@@ -0,0 +1,1685 @@
+
+  layoutItem->mLegendFilterByMap = itemElem.attribute( QStringLiteral( "legendFilterByMap" ), QStringLiteral( "0" ) ).toInt();
+
+  /* TODO: atlas
```
Anything blocking this now?

nyalldawson commented on tests/src/core/testqgscompositionconverter.cpp:
```diff
@@ -0,0 +1,669 @@
+    QVERIFY( count > 0 );
+  }
+  // TODO: attr table (not yet imported)
+
```
Is this to do outdated?

nyalldawson commented:
Looking good - feel free to merge when the comments are addressed (and Travis passes)

elpaso committed:
[layout] Added template method to retrieve QgsLayoutObjects

elpaso committed:
[layout] Remove legacy import

elpaso committed:
[layout] Update reference images

elpaso committed:
[layout] Address PR comments and add test for layoutObjects template

elpaso commented:
Server tests are failing due to missing layouts. I guess I need to fix the server too.

elpaso commented:
...  or better: temporarily disable the tests, merge this PR and then fix all the server stuff altogether in a separate PR and re-enable the tests when done.

@nyalldawson opinions?

nyalldawson commented:
+1 to disable and merge, given that we'll be working on the server transition next and it's only a temporary disable.

elpaso committed:
[layout] Temporarily disable WMS server tests

Until the composition -> layout transition is done

nyalldawson commented:
Great work - thank you!!

nirvn commented:
@elpaso , wouhou!
